### PR TITLE
[JIT] Add Exit Transform / Convert To SSA to docs

### PR DESCRIPTION
**Pull request by eellison**

### CODE REVIEW

File: torch/csrc/jit/docs/OVERVIEW.md [hunk 1]
```diff
@@ -548,7 +549,97 @@ This makes it possible to use most of the compiler functionality when python is
...
+
+```python
+if i < 0:
+  raise Exception("Negative input")
+else:
+  return math.sqrt(i)
+print(i)
+```
+
+We initially considering doing the Transform pass before Loads and Stores were removed from the graph. However, this breaks when a loop carried variable
```
Comment (Krovatkin) [inline torch/csrc/jit/docs/OVERVIEW.md]: consider**ed**?

File: torch/csrc/jit/docs/OVERVIEW.md [hunk 2]
```diff
@@ -548,7 +549,97 @@ This makes it possible to use most of the compiler functionality when python is
...
+
+## Conversion To SSA ##
+
+[script/convert_to_ssa.cpp](../script/convert_to_ssa.cpp)
+
+Converting to SSA works in multiple parts.
+- First, we add loads and stores to control flow operators (ifs & loops).
+- Then we erase Break & Continue statements from the graph and replace them with `prim::LoopContinuation`. `prim::LoopContinuation` has the form `LoopContinuation(%loop_continue_condition, %loop_carried_vars)`. Break Statements have the continue condition set to false, and Continue statements inline the loop condition. %loop_carried_vars are the loop carried variables of the inner most loop that contains the Break or Continue statement, are added by inserting prim::Loads calls at the location of the statement.
+- Then we inline the loop condition into the graph loops.
```
Comment (ZolotukhinM) [inline torch/csrc/jit/docs/OVERVIEW.md]: What is graph loops? :)

File: torch/csrc/jit/docs/OVERVIEW.md [hunk 3]
```diff
@@ -548,7 +549,97 @@ This makes it possible to use most of the compiler functionality when python is
...
+- Next we erase loads and stores, removing all Stores and replacing all loads
+with whatever the in-scope value of the variable name is.
+- Finally, we remove `prim::LoopContinuations` and `prim::ReturnStmts` in the exit_transform pass.
```
Comment (ZolotukhinM) [inline torch/csrc/jit/docs/OVERVIEW.md]: Nit: `prim::LoopContinuation`s

File: torch/csrc/jit/docs/OVERVIEW.md (hunk 1 above)
Comment (ZolotukhinM) [inline torch/csrc/jit/docs/OVERVIEW.md]: Nit: considered

File: torch/csrc/jit/docs/OVERVIEW.md [hunk 4]
```diff
@@ -548,7 +549,97 @@ This makes it possible to use most of the compiler functionality when python is
...
 
-The Environment object tracks the assignment of variable names to SugaredValues during compilation. It is local to the compiler file. A stack of environments exist, with a new environment being created for sub-blocks introduced by control flow. The Environment also handles turning the AST representation into SSA-form by tracking which variables were modified inside a sub-block and inserting the correct inputs/outputs to the Blocks of if-statements and loops.
+The Environment object tracks the assignment of variable names during compilation. It is local to the compiler file. A stack of environments exist, with a new environment being created for sub-blocks introduced by control flow. The Environment keeps two tables, one for values which are not first class in the type system (Sugared values) and a type table for values which are. When first class values are set, we emit a prim::Store, and when they are referenced we emit a prim::Load. Sugared values are not re-assignable. The graph is converted to SSA in the convertToSSA pass.
+
+## Conversion To SSA ##
+
+[script/convert_to_ssa.cpp](../script/convert_to_ssa.cpp)
+
+Converting to SSA works in multiple parts.
```
Comment (ZolotukhinM) [inline torch/csrc/jit/docs/OVERVIEW.md]: Could you please add a high level overview of why we're doing this (in contrast to "what")? It might be not obvious why this all is necessary and how it relates to classic SSA. For instance, in the conventional definition of SSA the central part is phi-nodes. We don't have them because we work only with structured control flow, but we never mentioned that here. But that's what explains all of what's happening here: we're trying to represent constructs with breaks/continues/returns in a form of well-structured CFG that only have loops with no early exits and ifs.
Another thing that might be confusing is the load/store part - we're not loading or storing anything there, it's just a trick we use to figure out input/output values of a block, and I think it's also worth mentioning somewhere.

File: torch/csrc/jit/docs/OVERVIEW.md [hunk 5]
```diff
@@ -548,7 +549,97 @@ This makes it possible to use most of the compiler functionality when python is
...
+with whatever the in-scope value of the variable name is.
+- Finally, we remove `prim::LoopContinuations` and `prim::ReturnStmts` in the exit_transform pass.  
+
+## Exit Transform ##
+
+[script/exit_transform.cpp](../script/exit_transform.cpp)
+
+This pass takes in a graph where LoopContinuation & ReturnStmts exist in the graph and erases them, correctly setting block outputs. `prim::LoopContinuation(*vals)` means that the values are targeting the most recent loop block. `prim::ReturnStmt(*vals)` means that the values are targeting the most recent Closure or Graph Block.
+
+Once we hit an Exit Node, we do not not execute any further instructions until the exit target has been reached. If we encounter a node that contains nested blocks that may have hit an exit node, such as an if statement that exits in one block and does not exit in the other, we use a boolean value to indicate if the exit has been hit or not. Then, we conditionalize further execution.
```
Comment (ZolotukhinM) [inline torch/csrc/jit/docs/OVERVIEW.md]: Nit: "Once we hit an Exit Node" sounds a bit strange to me as we don't execute anything here. Maybe rephrase it a bit (like "if a block has an exit node, further instructions will never be executed").

**Review by ZolotukhinM — APPROVED** [review]
Thanks for writing it up! It's really important to have such documents, it greatly helps to understand how stuff works.

Please find some comments inline and please fix failing tests before landing!

File: torch/csrc/jit/docs/OVERVIEW.md [hunk 6]
```diff
@@ -548,7 +549,97 @@ This makes it possible to use most of the compiler functionality when python is
...
+## Conversion To SSA ##
+
+[script/convert_to_ssa.cpp](../script/convert_to_ssa.cpp)
+
+Converting to SSA works in multiple parts.
+- First, we add loads and stores to control flow operators (ifs & loops).
+- Then we erase Break & Continue statements from the graph and replace them with `prim::LoopContinuation`. `prim::LoopContinuation` has the form `LoopContinuation(%loop_continue_condition, %loop_carried_vars)`. Break Statements have the continue condition set to false, and Continue statements inline the loop condition. %loop_carried_vars are the loop carried variables of the inner most loop that contains the Break or Continue statement, are added by inserting prim::Loads calls at the location of the statement.
+- Then we inline the loop condition into the graph loops.
+- Next we erase loads and stores, removing all Stores and replacing all loads
+with whatever the in-scope value of the variable name is.
```
Comment (Krovatkin) [inline torch/csrc/jit/docs/OVERVIEW.md]: this seems to be a little bit too dry. How are we tracking in-scope values? Are we doing something similar to what we used to do in `compiler.cpp` before? Ideally, we could use a simple example that shows a graph with loads and stores and then the one w/o it right before we remove `prim::LoopContinuations`.

File: torch/csrc/jit/docs/OVERVIEW.md [hunk 7]
```diff
@@ -548,7 +549,97 @@ This makes it possible to use most of the compiler functionality when python is
...
+
+## Conversion To SSA ##
+
+[script/convert_to_ssa.cpp](../script/convert_to_ssa.cpp)
+
+Converting to SSA works in multiple parts.
+- First, we add loads and stores to control flow operators (ifs & loops).
+- Then we erase Break & Continue statements from the graph and replace them with `prim::LoopContinuation`. `prim::LoopContinuation` has the form `LoopContinuation(%loop_continue_condition, %loop_carried_vars)`. Break Statements have the continue condition set to false, and Continue statements inline the loop condition. %loop_carried_vars are the loop carried variables of the inner most loop that contains the Break or Continue statement, are added by inserting prim::Loads calls at the location of the statement.
```
Comment (Krovatkin) [inline torch/csrc/jit/docs/OVERVIEW.md]: maybe, we should add why we decided to start delaying converting to SSA?

File: torch/csrc/jit/docs/OVERVIEW.md (hunk 5 above)
Comment (Krovatkin) [inline torch/csrc/jit/docs/OVERVIEW.md]: What's an Exit Node? `prim:: LoopContinuations` and/or `prim::ReturnStmts`?

File: torch/csrc/jit/docs/OVERVIEW.md [hunk 8]
```diff
@@ -548,7 +549,97 @@ This makes it possible to use most of the compiler functionality when python is
...
+```
+
+The pass also keeps track of nodes or blocks that will always throw Exceptions so that we do not unnecessarily conditionalize execution. In the following example, we can treat the if statement as always Returning and remove the `print` statement.
+
+```python
+if i < 0:
+  raise Exception("Negative input")
+else:
+  return math.sqrt(i)
+print(i)
```
Comment (Krovatkin) [inline torch/csrc/jit/docs/OVERVIEW.md]: nitpick: we could a comment `# unreachable code`

File: torch/csrc/jit/docs/OVERVIEW.md (hunk 5 above)
Comment (Krovatkin) [inline torch/csrc/jit/docs/OVERVIEW.md]: > If we encounter a node that contains nested blocks that may have hit an exit node, such as an if statement that exits in one block and does not exit in the other, we use a boolean value to indicate if the exit has been hit or not.

To double check I parsed it correctly, this means that all outer blocks that contain the exit node that was hit will now include a test to check if the exit node was triggered?

File: torch/csrc/jit/docs/OVERVIEW.md [hunk 9]
```diff
@@ -548,7 +549,97 @@ This makes it possible to use most of the compiler functionality when python is
...
+```
+
+-> transforms to
+
+```python
+continue_loop = i < 5
+while continue_loop:
+  if i == 3:
+    i = i + 1
+    continue_loop = i < 5
```
Comment (Krovatkin) [inline torch/csrc/jit/docs/OVERVIEW.md]: `continue_loop` is inserted at the end of every inner block?

**Review by Krovatkin — COMMENTED** [review]
Thanks for writing this up! It looks great! I added a few comments
